### ISSUE TITLE
Multipart Formula

### ISSUE THREAD

**Issue by matthieugomez** [JuliaStats/StatsModels.jl#21]
Beyond a `formula`, a `fit` function typically has several arguments that may refer to dataframe variables. To give some examples, there can be additional symbolic terms for models more complicated than GLM, variables corresponding to standard errors computations, weight variables, or logical expressions that refer to variable names to subset a dataframe (I use all these arguments in   (https://github.com/matthieugomez/FixedEffectModels.jl)[FixedEffectModels]).

The recent move from `@~` to `@formula` is nice because it recognizes that formulas are not particular: a lot of other arguments of `fit` also require to capture variable names.

It would be nice to think collectively about the best syntax to `fit` a model. I have thought about the following possibilities:

```julia
# syntax 1: define a different macro for each argument that needs to capture variable names
fit(df, @formula(y ~ x1), @weight(x2), @vcov(cluster(x3)), @where(x4 >= 0)))

# syntax 2: define multipart formula `@model(expr, kwargs...)`. 
fit(df, @model(y ~ x1, weight = x2, vcov = cluster(x3), where = x4 >= 0)))

#syntax 3: define a `@fit(expr1, expr2, kwargs)` macro (syntax closest to [Stata](http://www.stata.com/manuals13/rregress.pdf) and to [DataFramesMeta.jl](https://github.com/JuliaStats/DataFramesMeta.jl))
@fit df y ~ x1 weight = x2 vcov = cluster(x3) where = (x4 >= 0))
@fit(df, y ~ x1, weight = x2, vcov = cluster(x3), where = (x4 >= 0))
```

I don't really have a preference for one syntax in particular. I just think it would be nice to decide on one. An additional benefit is that such an "approved" syntax would standardize the names of commonly used arguments like "weight" "vcov" "where" across different packages. This is the way it is in Stata and I think it leads to a more user experience when switching between different estimations.

**Comment by ararslan** [comment]
I don't understand the use case for capturing variable names anywhere but in the actual construction of the model terms. Also, macros don't support keyword arguments.

**Comment by matthieugomez** [comment]
I have three difference use cases in  [FixedEffectModels](https://github.com/matthieugomez/FixedEffectModels.jl)), have a look at the Readme. Or look at all the R packages that rely on [multipart formula](https://cran.r-project.org/web/packages/Formula/index.html)
Macros can work with keyword arguments as they are just particular expressions. For instance `@test` syntax is
```
using Base.test
@test 1 ≈ 0.99 atol = 1e-2
```

**Comment by nalimilan** [comment]
I always thought passing symbols for variable names would be enough. That doesn't work for `where`, but I'm not sure I like the idea of supporting `where` in model fitting functions, as it sounds simpler to perform this step separately.

**Comment by matthieugomez** [comment]
The examples I have talked about are the simplest use cases that don't require to talk about specific estimation method. But, if you are not convinced by these simple examples, let me delve into [FixedEffectModels](https://github.com/matthieugomez/FixedEffectModels.jl). An estimation command requires the user to specify (i) a formula (ii) a set of high dimensional fixed effects (iii)  set of variables to cluster on.
A syntax that use symbol for everything except the formula looks clumsy:
```julia
fit(@formula(y~x1+x2), fe = (:x3, :x4), vcov = (:cluster, :x5, :x6))
```
Rather than this syntax,  package developoers in R have preferred to use [multipart formula](https://cran.r-project.org/web/packages/Formula/index.html). For instance `felm` uses the syntax:
```julia
lfe(y~x1+x2 | x3 + x4 | x6 + x7)
```
The issue with this solution is that it is not clear which part of the formula refers to. This is why I am pushing for a keyword syntax, something similar to
```julia
@model(y~x1+x2, fe = x3 + x4, vcov = cluster(x6 + x7))
```

**Comment by nalimilan** [comment]
I guess what we could do is that a general macro like `@model` or `@fit` would transform every expression starting with `~` into a formula, so that you don't need to repeat `@formula`. But automatically transforming `fe = x3 + x4` into a formula sounds difficult/impossible to me: it doesn't generalize, and yet that macro would have to be defined only once in StatsModels.

So the solutions I can imagine are like:
```julia
@fit(y ~ x1+x2, fe = (:x3, :x4), vcov = cluster(:x5, :x6))
@fit(y ~ x1+x2, fe = ~ x3 + x4, vcov = cluster(~ x5 + x6))
```

**Comment by matthieugomez** [comment]
I don't think this overloading would be a good idea. This overloading would completely abuse mathematical notations : clustered variables refer to methods to compute standard errors, not variables to regress on. Moreover,  what if there are some functions defined `fe` or `cluster` in the namespace? 
It is already possible to have function calls in R, yet it did not prevent people to require multipart formulas.

**Comment by nalimilan** [comment]
So, would one of my examples suit your needs?

**Comment by ararslan** [comment]
`fit` wouldn't even need to be a macro for the former example.
```julia
fit(@formula(y ~ x1 + x2), fe=(:x3, :x4), vcov=cluster(:x5, :x6), ...)
```
You just need a `Symbol` varargs method for `cluster`.

**Comment by matthieugomez** [comment]
Yes, that would work. Another solution could be to simply capture all the arguments, rather than the one prefixed by `~`, no? (as in syntax 3). Each package can then define the right `fit` function. Something like
```julia
macro fit(x, args...)
	:(fit(@formula($(esc(x))), $args...))
end 
```

**Comment by nalimilan** [comment]
But not all arguments must be formulas. And in some cases you explicitly need to refer to local objects rather than to data frame columns.

**Comment by andreasnoack** [comment]
What about just using several `@formula` arguments?

**Comment by matthieugomez** [comment]
@nalimilan I agree, but this problem is common to every dplyr-like data manipulations (like DataFramesMeta).

@andreasnoack Something like this would work indeed.
```julia
fit(df, @formula(y ~ x1), fe = @formula(x3&x4 + x4), weight = @formula(x5), vcov = cluster(@formula(x3+x4))
```
It is quite verbose.

**Comment by matthieugomez** [comment]
Maybe we can come back to this issue once there is a commonly accepted way to handle expressions that may or may not use variable names (i.e. something akin to a dplyr-like package). I just wanted to have this issue somewhere — I think it would be great to nail the syntax of estimation commands.

**Comment by nalimilan** [comment]
I don't think the two issues are completely related (and that's good news since it makes it easier to fix the present one). Query macros need to deal with identifiers which most frequently refer to data frame columns, with a way to escape this occasionally (e.g. using `$x`, which was evoked at some point). `fit`/`@fit` is quite different, since most arguments do *not* refer to columns: data frame object, distribution family, number of iterations, etc. Arguments which have to refer to columns are more the exception than the rule. What we need is a way to make them not too painful to use, but not the default either. I'd say my proposals are reasonable compromises, though there are other solutions.

**Comment by matthieugomez** [comment]
Right, I agree with this.

**Comment by Nosferican** [comment]
I am working on a econometrics package and would be nice for implementing estimators for panel data (cross sectional unit, time unit, cluster, etc.).

**Comment by nalimilan** [comment]
@Nosferican Can you give some illustrations of your needs like @matthieugomez did above?

**Comment by Nosferican** [comment]
It would be quite similar.

- Model specification as formula
- DataTable / DataFrame
- Indicators for cross sectional unit and time indicator (x, t)
- Instruments can be closely related to the formula argument
- Clustering for standard errors (one or more features)
- Weights for observations

```
fit(@formula(y ~ x1 + x2 + x3), df, @xt(x, t), family, link, estimator,
    @instruments(x3 ~ z1 + z2), @cluster(x4), @weights(x5))
```

Another examples of packages using expanded syntax in the formulas other than [FixedEffectModels.jl](https://github.com/matthieugomez/FixedEffectModels.jl) is the [MixedModels.jl](https://github.com/dmbates/MixedModels.jl)

```
m = fit!(lmm(Yield ~ 1 + (1 | Batch), ds))
m2 = fit!(lmm(y ~ 1 + dept*service + (1|s) + (1|d), inst))
fm3 = fit!(lmm(Reaction ~ 1 + Days + (1+Days|Subject), slp))
```

**Comment by nalimilan** [comment]
OK, looks similar indeed.

I think the design should be as generic as possible, i.e. it shouldn't require particular models to define macros like `@cluster` or `@weights`, since these are going to clash between packages. So arguments referring to variables inside the data frame should either be formulas, or (a tuple of) symbols when a formula wouldn't be natural (e.g. `@xt(x, t)` can be `xt=(:x, :t)`, or even `x=:x, t=:t`).

To avoid the tedious requirement repeat `@formula`, the simplest and most systematic solution would be to implement a `@fit` macro which would replace all arguments starting with `~` with a formula. People would still be able to use `@formula` explicitly with the `fit` function if they prefer. Any objections to that approach?

(I would say the case of MixedModels is slightly different, since I don't think @dmbates has any complaints about the `(1|x)` syntax: these terms are really part of the same equation, contrary to clusters or instruments.)

**Comment by matthieugomez** [comment]
@nalimilan Your solution would work but I don't really like this syntax.  The whole point of `@formula` is to make `~` as a prefix to capture an expression disappear. It is is an R quirk and I would be glad to have it go away.

What about Syntax 2? In this syntax, `@formula` accepts multiple arguments, that are be captured. The captured arguments are then passed to a function `_formula`, which can be imported by each package.

```
# definition in StatsModels
using StatsModels
macro formula(args...)
    Expr(:call, :_formula, (transform_expr(a) for a in args)...)
end
function transform_expr(expr)
   if isa(expr, Expr) && expr.head == :(=)
      return Expr(:kw, expr.args[1], Base.Meta.quot(expr.args[2]))
   else
      return Base.Meta.quot(expr)
  end
end
function _formula(ex)
       StatsModels.Formula(ex.args[1], ex.args[2])
end

# Definition in my package
function _formula(ex::Union{Expr, Symbol}; fe::Union{Expr, Symbol} = :nothing, weight::Union{Expr, Symbol} = :nothing, vcov::Union{Expr, Symbol} = :nothing)
  println("it works")
end
@formula(y ~ x1 + x2 + x3, fe = x1 + x2, weight = x3, vcov = cluster(x4 + x5))
```

**Comment by nalimilan** [comment]
The problem with that syntax is that it puts inside a common `@formula` call arguments which are unrelated: `weights` has nothing to do with `vcov` (or at least not more than with `maxiter`). I agree that `~` is a bit annoying when there is not LHS, but that's a small issue in comparison. Also, that symbol is not only used to escape an expression (which can be done using `:()` in Julia), it also indicates that the argument uses the formula syntax, i.e. terms and interactions can be specified using `+`, `&` and `*`. To me, that's a good reason to use the same symbol everywhere. If an argument doesn't need this syntax, it should just be a (list of) symbol(s).

**Comment by Nosferican** [comment]
I believe the difference would most likely be whether we want `@formula` to be specific to a model specification with form `response ~ variables` with variables allowed to perform contrasts specified through `+`, `*`, and `&` or generalize it. Stata is a more pure formula system which operates with variables contrasts versus R which has enhanced formulas for on-the-fly feature generation and multipart formulas (i.e., it has a single lhs and the rhs is an array divided by the `|` operator). Examples include Stata `i.`, `c.`, `#`, `##`, `L`, and `D` operators and R's `log`, `poly`, `I`, `*`, and multipart features.

Stata: `reg y X1 i.X2 c.X3 X4#X5 X6##X7 L.X8 D.X9, cl(X10)`
R: `log(y) ~ poly(X1, 2) + I(X2 >= 2) + X3 + X4 * X5 | X3 ~ Z1 + Z2`

While I personally prefer to compute the features outside the formula in certain cases such as when computing the margins of a logistic regression (logit, multinomial, or ordinal) it is essential to keep track which variables are related and how which is done through passing the transformation in the formula. Easier cases such as with instrumental variables can be traced, but for harder cases it is almost impossible unless the feature engineering occurs in the formula. Other times it is just for convenience (e.g., weights as a function of the residuals).

While R uses the `|` operator to create multi-part rhs, I am a fan of keywords (pray every night for Julia to get keyword argument mapping like in R). I believe a Stata approach of a core expanded formula should be followed by `,` keyword arguments.

Guess my vote is in favor of keeping the formula with the bare minimum (indispensable and related) and pass additional arguments as keywords. For example, `StatsModels.ModelFrame` uses the formula, but passes the contrast dictionary as an additional argument.

**Comment by matthieugomez** [comment]
@nalimilan I see what you mean. However, another issue with your solution is that it defines 2 different syntaxes to refer to variable names: symbols for arguments that always refer to only one variable (like `weight`),  formula for arguments that may refer to multiple variable names (like `fe`). I think the resulting syntax looks clumsy
```julia
@fit(y~x1, fe = ~x2, weight = :w)
```

Finally, I disagree when you say that `weights` has nothing to do with `vcov`. All arguments inside `@formula`, i.e. the ones with variable names, tend to correspond to the model specification (i.e. weight, how to compute standard erorrs, etc). In some sense the macro `@formula` could be called `@model`.

**Comment by nalimilan** [comment]
> Or it would require to impose people to use ~varname rather than :varname.

I guess we could also allow both if it turns out to be confusing.

> Your arguments agains Syntax 2 are well taken, although I tend to disagree when you say that weights has nothing to do with vcov (compared to maxiter). All arguments inside @formula (i.e. all the arguments that refer to variable names in Syntax 2), tend to correspond to the model specification (i.e. weight, how to compute standard errors, etc). In some sense the macro @formula could be called @model.

Maybe they "tend to", but that's not always the case. For example, the model family for GLMs is clearly related to the model specification, but it's not a formula. And we haven't considered more exotic models yet.

> Now, if you really don't like the idea of splitting the arguments between @formula and fit (i.e. Syntax 2), we can still use something like Syntax 3, i.e. a macro @fit that captures all arguments. Each developer can then deal with evaluating each argument separately depending on the keyword argument

I don't think that works either. The macro needs to decide whether to transform arguments *before* it knows the type of the model, so it cannot use dispatch to let each package choose a behavior.

**Comment by matthieugomez** [comment]
The macro would just capture all the expressions and and pass them to the function. the function would then specify which argument should be evaluated. Is there a problem with that?

**Comment by nalimilan** [comment]
Yes, the function would receive expressions which refer to a different scope. For example, `maxiter=n` would be transformed into `maxiter=:n`, but `n` wouldn't exist in the function's scope (actually, the scope of the module), or it could be shadowed by a local variable, by another keyword argument... More generally, calling `eval` inside functions isn't recommended in Julia.

**Comment by matthieugomez** [comment]
Correct me if I'm wrong but I don't think this is a problem. If I understand how macro works, 
`@fit(fe = z, maxiter = n)` could just call `_fit(:(y~z), fe = :z, maxiter = :n)` which would return the expression `:(fit(@formula(y~x), fe = ~z, maxiter = n))`. This expression will then be evaluated in the environment.

Just to give a sketch:
```
macro fit(args...)
    _fit((transform_expr(a) for a in args)...)
end
function transform_expr(ex)
   if isa(ex, Expr) && ex.head == :(=)
      return Expr(:kw, ex.args[1], Base.Meta.quot(ex.args[2]))
   else
      return Base.Meta.quot(ex)
  end
end

function _fit(fe = :nothing, maxiter = :nothing)
   Expr(:call, :fit, Expr(:kw, :fe, Base.Meta.quot(fe)), Expr(:kw, :maxiter, maxiter))
end
```

**Comment by nalimilan** [comment]
Again, that's the same problem: your `_fit` function will conflict with the one defined in another package since its signature does not have any specific to your package.

**Comment by matthieugomez** [comment]
The keyword arguments are specific to my package. No?

**Comment by ararslan** [comment]
AFAIK Julia doesn't care about the names of the function arguments, just their types. So unless you have a function argument that accepts only a type defined in your package, it will conflict.

**Comment by nalimilan** [comment]
...and keyword arguments do not even participate to dispatch.

**Comment by matthieugomez** [comment]
Thanks for the precision. I have to think about it more but I can think of two other possibilities. One would be be to define a macro  instead of relying on a `@fit` defined in `StatsModel`.
```julia
@reg(y~x, fe = z, maxiter = 10)
```
Another possibility could be to define a specific symbol for keyword arguments that should be captured, say `=>`
```
@fit(formula => y~x, fe => x, maxiter = 10)
```

**Comment by Nosferican** [comment]
A big advantage of the macro approach over the symbols only is that it allows for operations such as contrasts (`&`, `&&`) and eventually feature engineering directly on the formula.
I would be happy to have `@model` which would parse a `@formula` and manage keyword arguments with a `=>` assign/capture operator. This approach might even help packages that rely heavily on generating a model which then is passed to a fit function.

**Comment by nalimilan** [comment]
Each package can of course provide its own macro, but it would be nice to be able to harmonise the model fitting interfaces. Regarding propositions 2 and 3, I really don't see the point of reinventing a syntax for formulas.

**Comment by ararslan** [comment]
Perhaps @dmbates could help unify the tribes.

**Comment by dmbates** [comment]
Gee, thanks, @ararslan 

I was trying to keep my head down.

**Comment by matthieugomez** [comment]
@nalimilan what kind of problem would it create to capture every arguments (except for, say, the dataframe),  imposing users to directly insert values in the AST, i.e.
```julia
n = 10
@fit(df, y~x, maxiter = $n)
```

**Comment by nalimilan** [comment]
The main problem would be that it's not standard so it would be surprising/confusing. Note that even `df` would have to be written as `$df` for that syntax to be fully consistent, and the same would apply e.g. to the model family (`$(Binomial())`). So you're trading `~` in formulas with `$` for other arguments, which is probably not a win.

**Comment by matthieugomez** [comment]
Thanks. To ensure the consistency problem you mention, what about a `fit(df, ModelType, @model)` syntax , where every argument in `model` is captured?
```julia
fit(df, LinearModel, @model(y~x, Binomial(), maxiter = 10))
```
One would only need to use the symbol `$` when using `@model` within a function i.e. during non interactive use.
```julia
function f(x, n)
   fit(df, LinearModel @model(y~x, $(x), $(n))
end
```
```

**Comment by nalimilan** [comment]
I don't like the idea of a syntax that only works in the global scope (and I suspect I'm not the only one).

**Comment by matthieugomez** [comment]
I have now defined the macro `@model` in FixedEffectModels [here](https://github.com/matthieugomez/FixedEffectModels.jl/blob/master/src/utils/model.jl). It accepts a formula and a set of keyword arguments, and returns a `Model` with the captured expressions:
```julia
using FixedEffectModels
@model y ~ x fe = v1 + v2 weights = w
# Formula: y ~ x
# fe: v1 + v2
# weights: w
```
I with StatsModels had a similar macro.

**Comment by lbittarello** [comment]
Latecomer to the party. I've been having the same issues, but I wasn't aware of this discussion. My solution consisted of separating data processing from estimation and using keywords. For example:

`data = Microdata(DF, response = "y", control = "1 + x1 + x2", treatment = "t", instrument = "z", weight = "w")`
`results = fit(IV, data, method = "TSLS")`

It seems similar to Syntax 2 above.

By the way, I don't like the `y ~ x` structure. I understand that it's understand, but it's awkward. Consider linear IV and IV weighting, for instance. In the first case, y is regressed on x and t, so the formula makes sense. In the second case, z is regressed on x to construct weights for a regression of y on t, so the formula doesn't make as much sense. They shouldn't need different inputs though...

**Comment by nalimilan** [comment]
> It seems similar to Syntax 2 above.

Well, not completely, since variables and formulas are passed as strings, so that `Microdata` is a standard function, not a macro. Since quotes make it possible to distinguish between formulas and standard arguments (like DF), this solution is more similar to my proposal, which is to use a macro and indicate that an argument is a formula via `~` (or possibly `:` for single variables).

> By the way, I don't like the y ~ x structure. I understand that it's understand, but it's awkward. Consider linear IV and IV weighting, for instance. In the first case, y is regressed on x and t, so the formula makes sense. In the second case, z is regressed on x to construct weights for a regression of y on t, so the formula doesn't make as much sense. They shouldn't need different inputs though...

Sorry, I'm not familiar with the example you are using, so I don't understand the problem. Could you elaborate? Note that you don't need to use `y ~ x` if that doesn't make sense for your use case: formulas are relatively flexible.

**Comment by Nosferican** [comment]
In my latest iteration I have decided to follow @matthieugomez scheme for the core formula:
```
response ~ exogenous + (endogenous ~ instruments)
```
for fixed effects a one-sided formula
```
absorb = fe1 + fe2 * fe3
```
which is similar to Stata's `reghdfe`.
For the "model" I am using a `@model` which also includes the data,
```
model = @model(data = df, formula = y ~ x1 + (z1 ~ Z1), absorb = x + t, weights = w)
```
`@model` captures the values and puts them in a dictionary which is passed to my struct generator which parses it and populates the package struct `<: StatsBase.RegressionModel`. `StatsBase.fit!` is expanded to have a method for the package struct.

**Comment by nalimilan** [comment]
I'm still not convinced that a macro capturing all arguments is a good idea. That means `@model` will only work for your package, and if any other package wants a similar mechanism it will have to use another name or experience conflicts. Also, as I noted, it doesn't allow passing references to local variables (except via ugly hacks for the global scope only). Really, requiring people to write `~ x  + t` and `~w` or `:w` doesn't sound like a big deal and it would make the macro much more standard.

**Comment by Nosferican** [comment]
A possible fix for not having to have a qualified `@model` is to have it part of `StatsModels`, or `DataFrames` for the moment, as:
```
macro model(args...)
	args = Dict(StatsModels.parsemodelargs.(args))
	model = StatsModels.modelbuilder(args)
	return model
end
function StatsModels.modelbuilder(args::Dict{Symbol,Any})
        # Some code that dispatches the module's modelbuilder based on
        # get(args, :module, :Default)
end
StatsBase.fit!(model::ModuleStruct)
```
and dispatch on the module the user wants to use.
The only argument I can think of that could or should be local I think is the data. I use,
```
function parsemodelarg(obj::Expr)
    args = getfield(obj, :args)
    field = first(args)
    value = last(args)
    if field == :data
        tmp = Expr(:call, :(x -> return x), :data)
	tmp2 = getfield(tmp, :args)
	tmp2[2] = value
	setfield!(tmp, :args, tmp2)
	value = eval(Main, tmp)
    end
    ...
    return (field, value)
end
```
which could potentially allow for accessing that local reference in other spaces besides the Main module space (global). Again, I that would only be used in very limited instances I would think, but unsure if that can be used in all applications that you can think of. Otherwise, one can keep data apart and do something like `FixedEffectModels.@model`
```
@reg(df, @model(...))
```
A limitation for using `~ x + t` is that `@formula` and `ModelFrame` expect a `:lhs` that is not `nothing` in order to work limiting its use.

**Comment by nalimilan** [comment]
Anything that depends on `eval` is a no-go for any standard package like StatsModels (and it's not a great idea in general). And you can't know in advance which argument is going to be a formula, and which one should be a local variable -- except by hardcoding their list in StatsModels, but we don't want to restrict what packages can do.

> A limitation for using ~ x + t is that @formula and ModelFrame expect a :lhs that is not nothing in order to work limiting its use.

Maybe we can work on finding a good solution with formulas? We could allow an empty LHS, or use `:(...)`, or find another symbol...

**Comment by kleinschmidt** [comment]
something like `@model df GLM(z ~ 1+x*y, family=:binomial)`

**Comment by ararslan** [comment]
I believe @kleinschmidt has implemented one-sided formulas, e.g. `@formula(~ x + t)`, in a branch somewhere.

**Comment by ararslan** [comment]
I hesitate to get too far into macro purgatory where it becomes unclear what things are doing.

**Comment by kleinschmidt** [comment]
I think it's the only way to deal with this though, right?

**Comment by Nosferican** [comment]
Basically,
```
abstract type ModelBuilder end
macro model(data::DataFrames.AbstractDataFrame, mb::ModelBuilder)
    # Packages will create a model builder to use with the @model in order to parse
    # arguments and return their PkgStruct <: StatsBase.RegressionModel for instance with
    # to be used in StatsBase.fit!(model::PkgStruct)
end
```
Formulas will allow for one sided for those that require it such as absorbing variables.
Maybe expand ModelFrame, ModelMatrix to take multiple formulas? That way it could fix the issues of missing values and return a list of ModelFrames, ModelMatrix or something?

Brainstorm,
```
mypart = mymodelbuilder(formula = @formula(y ~ x1 + x2 & x3),
                                            iv = @formula(z1 ~ z2 + z3),
                                            absorb = @formula(~ fe1 + fe))
model = StatsModels.@model df mypart
StatsBase.fit!(model)
StatsBase.coeftable(model)
```

**Comment by nalimilan** [comment]
The point of using a macro like `@model` is that it would transform all formula argument (identified by their `~`) into `Formula` objects, which can then be treated like normal objects. So you wouldn't need to specify `@formula` for each argument (you don't need a second macro like `@model` if you do that).

**Comment by Nosferican** [comment]
For correctly parsing the `Expr` with `ex.head == :(~)` it would need to work with:
1. Standard (lhs::Symbol, rhs::Union{Expr,Symbol,Integer,Void}) response ~ exogenous
2. Instrumental (lhs::Union{Symbol,Expr}, rhs::Union{Symbol,Expr}) endogenous ~ additional instruments ... maybe include here those à la [MixedModels](https://github.com/dmbates/MixedModels.jl)?
3. One sided such as fixed effects (lhs::Void, rhs::Union{Symbol,Expr})

Arguments without `ex.head == :(~)` be passed as `Expr`?

**Comment by nalimilan** [comment]
Anything with `ex.head == :(~)` would be treated as if it was inside a `@formula` call. Then we can discuss what kind of formulas should be supported. Anything else would be treated as a standard argument to a standard function.

**Comment by kleinschmidt** [comment]
My half-baked proposal is that the `@model` macro is called with a data frame (or other tabular source) and a model constructor expression, and lowers to a container type kind of like `DataFrameStatisticalModel`, but which extracts all the arguments of the constructor that involve `~`, replacing them in the constructor call with model matrices (and storing the parsed formula data structures themselves).  To be concrete, that would look like

```julia
@model df GLM(y~x, family=Binomial)
# becomes
let formulas=(Formula(:x, :y), )
    TableModel{GLM}(df, formulas, GLM(modelmatrix(formulas[1], df), family=Binomial)
end
```

The `@model` macro and the `TableModel` type are defined in StatsModels, as are generic `fit!` methods for `TableModel{<:Any}` that just call `fit!` on the model field.  If you want to define special `fit!` behavior that does something other than just use the model matrices generated by the formulas, then you can defined depend on StatsModels and define methods for `fit!(::TableModel{MyModelType})` (and other methods from statsbase etc.).

It might be necessary to add an intermediate call to some kind of builder function to keep from eagerly evaluating the model matrices here in case packages want to do something weird with the formulas, again using a default fallback defined in StatsModels that does that step.

This would all require a bit of re-structuring of the whole API, but I think that's for the best.  I think the most disruptive thing is making sure everyone follows a two-step model construction and fitting API.

**Comment by kleinschmidt** [comment]
This might also require storing kw arguments with formulas in a dict in the `TableModel` for this to work for use cases where the kw arguments need special behavior.

**Comment by Nosferican** [comment]
Considering the current implementation I would actually favor establishing the different components of the model universally. For example:
```
abstract Formulae end
mutable struct EnhancedFormula <: Formulae
    response
    exogenous
    endogenous
    instruments
    fixedeffects
    weights
    panelid
    temporalid
    EnhancedFormula() = new()
end
```
Then `ModelFrame` / `ModelMatrix` for example can generate the model arrays (`y`, `X`, `z`, `Z`, `D`, etc.). For example, `coefnames` could be a `Dict{Symbol,Vector{String}}`. The basic macro would parse an enhanced syntax formula into a `EnhancedFormula`.
```
y ~ x1 * x2 + x3 + (z1 + z2 ~ Z1 + Z2) + absorb(fe1 + fe2) +
    weights(weight) + cluster(PID + TID) + PID(County) + TID(Year)
```
One can overload the `Base.show` to display it in a more conventional format for example.
If anyone wants to give it a try I could implement it.

**Comment by lbittarello** [comment]
I find that `EnhancedFormula` is similar to `Microdata`, but it is less flexible.

**Comment by kleinschmidt** [comment]
I'm really opposed to that, and I think it actually shows why the current API should be changed: it's too inflexible.  I don't want to have to specify every single possible additional part of the formula in StatsModels.  I'd much rather tear things down and rebuild them to be flexible enough that people can extend the as they like.

**Comment by Nosferican** [comment]
How is it restrictive? Right now I can accomplish the same and with virtually anything I want hacking away the current implementations... For example, for fixed effects I just `deepcopy` and `setfield!` the left hand side to the response in order to use them in `ModelFrame`, `ModelMatrix`... For iv I construct my formula and switch sides for endogenous and give it a symbol in the left for instruments. The idea with enhanced formula is to make it easier rather than each in their package to have to hack away to build what they want. The struct is modeled as incomplete so if one doesn't need a field one can leave it undefined.

**Comment by kleinschmidt** [comment]
I see the formula DSL as a way of describing how columns in a tabular data store should be converted to numeric columns for modeling.  How they are interpreted is up to the particular model and should (IMO) be left out of the specification of formulae themselves.

An important thing that is maybe missing from this discussion is that we're planning to change the formula DSL to allow for arbitrary functions to be passed through; this would allow for special functions to be defined in packages that need them as well, which would obviate the need for adding them to the formula DSL.

**Comment by nalimilan** [comment]
It's restrictive because you need to list in advance all possible fields a model might need. You included `temporalid` there, but if you weren't in this discussion we would never have guessed somebody might need it one day. What will we do once somebody tries to implement a new kind of model? We don't want to make a new StatsModels release every time somebody needs an additional field.

**Comment by Nosferican** [comment]
An alternative is to have just the various archetypes with few assumptions. The current model assumes only one type of formula (lhs::Symbol, rhs::Union{Expr,Symbol,Integer}). Enhanced formula is an child of the abstract `Formulae` and `ModelFrame` / `ModelMatrix` would be defined for `Formulae` so if a package has some needs outside that scope or is better served by another structure they just need to create a struct with their fields as a child of `Formulae`. The two alternatives are to do minimum (current `Formula` and allow to hack it to achieve what someone needs) or allow for more advanced use giving up some flexibility. For example, instead of a struct it could be a dictionary with `Formula` (of different types: standard, one sided, multiple left) and just have `ModelFrame` "correctly parse" them. Meaning the dataframe will keep all variables that appear in the union of the formulas, and create the appropriate `Vector{ModelFrame}`.

**Comment by Nosferican** [comment]
@kleinschmidt, regarding your approach, what about
in `StatsModels`
```
abstract type ModelSpecs end
modelbuilder(data::DataFrames.AbstractDataFrame, modelspecs::ModelSpecs) = 
        error("modelbuilder is not defined for $(typeof(modelspecs)).")
```
In the different packages,
```
mutable struct MyPkgModel <: StatsBase.RegressionModel
    # Some code
end
function StatsBase.fit!(obj::MyPkgModel)
    # Some code that fits the model
end
function StatsBase.coeftable(obj::MyPkgModel)
    # Some code that returns the summary of the model
end
mutable struct MyPkgModelSpecs <: StatsModels.ModelSpecs
    # Some code
end
function modelbuilder(data::DataFrames.AbstractDataFrames, modelspecs::MyPkgModelSpecs)
    # Some code that uses the data and MyPkgModelSpecs to return a MyPkgModel
end
macro @myparser(args...)
    # Code to parse my model specifications
end
```
API (WLOG let `@myparser` in this case be `@econmodel`)
```
modelspecs = @econmodel(formula = y ~ x + (z1 ~ Z),
        estimator = FD, vcov = cluster(PID + TID))
model = modelbuilder(df, modelspecs)
fit!(model)
coeftable(model)
```

**Comment by kleinschmidt** [comment]
Yeah, that's something like what I meant by an "intermediate model building stage".  The `@model` macro might lower to a constructor for `ModelBuilder{MyModel}(...)`, which would have a default for `ModelBuilder{<:Any}` in StatsModels but could be overloaded if package authors wanted to do something fancy with the formulas captured by the macro.

One improvement to the formula language is something I've started to play around with in StreamModels, using `Term` types to represent each term in the formula.  Then a formula itself could be a `FormulaTerm`, meaning you could (in principle) nest "formulas" within each other.  So something like `y ~ x + (z1 ~ Z)` would lower to `FormulaTerm(Placeholder(:y), MatrixTerm(Placeholder(:x), FormulaTerm(Placeholder(:z1), Placeholder(:Z))))`.  Then you can do what you want to with this intermediate representation at the model builder stage.  The default would be to plug the data frame columns into the placeholders and spit out a response and model matrix for something like a GLM (in which case there would be an error when you have an embedded formula).

**Comment by nalimilan** [comment]
Though it would be nice to have a user-friendly API which would directly take a dataset and fit the model one it, in addition to a lower-level API which would allow constructing a `ModelSpecs`/`ModelBuilder` first and then call `fit` on it separately. Maybe the user-friendly version could be provided by each package, e.g. with `@econmodel`. The lower-level version would only use generic functions like `@model`, `modelbuilder` and `fit`, but passing a type inheriting from `StatsBase.StatisticalModel` for dispatch.

**Comment by Nosferican** [comment]
My impressions after my initial play with DataFrames / StatsModels for Julia 0.7 are very positive. `StatsModels.Formula` seems a lot more flexible. For example, now I can have
```
myfunc(@formula(y ~ x1 + x2 & x3 + (z1 + z2 ~ Z1 + Z2) ~ fe1 + fe2), df, ...)
```
No longer I have to provide the parser and I can easily get each element from the `StatsModels.Formula`. Not sure if possible, but I would like to have `|` work similarly to `~` à la R's multipart, but `~` works well for now. As for capturing model specifications, I would favor having a `StatsModels.@modelspecs` or `StatsModels.@modeloptions`. That way all packages can have their non-formulae options captured by a common API. `StatsModels.@modelspecs` could be defined like
```
function parsearg(obj::Expr)
    args = obj.args
    @assert (args[1] == :(=)) & (length(args) == 3) "Model options should have form: Argument = Value"
    (obj.args[1], obj.args[2])
end
macro @modelspecs(args...)
    Dict(parsearg.(args))
end
```
This way each package can have their function collect their formula, data, and model options using a common API. For example,
```
reg(@formula(y ~ x1 + (z1 + Z1)), df, @modelspecs(estimator = FirstDifference, vcov = HC1))
```
This also encourages packages from exporting structs that might overlap.

**Comment by kleinschmidt** [comment]
@Nosferican I've started to work on a prototype that's even more flexible in StreamModels. The first major difference that is relevant to this discussion is that the internal representation of formulae is as a tree of `Term` sub-types, with a `FormulaTerm` as the root (thus allowing sub-formulae).  I'm still working out the details but this should allow for greater extensibility of the syntax (e.g., a package could add special parsing rules for whatever calls they'd like).

The second major difference (which is even more preliminary) is that it introduces a `@model` macro that captures all the arguments to a normal model constructor call, converts formulae to Formula objects, and stores all the positional and keyword arguments in a `ModelBuilder` struct. The `ModelBuilder` type is parametrized by the type of the model (or more generally the call), which allows for overloading default behavior by specializing on the model/call type.  For instance, constructing an actual model happens by calling `build(::ModelBuilder, datasource)`, but you can easily add a specialized method `build(::ModelBuilder{MyModel}, datasource)` to implement whatever transformation fo the captured arguments (including formulae) that you want.  The default (as I imagine it) would be to just evaluate all the formulae into numerical matrices/vectors (along the lines of what `ModelMatrix(ModelFrame(f, df))` currently does), and plug them into the constructor in place of the formulae, along with the rest of the captured arguments.

I'm really keen to know what you think of that architecture, and whether it would allow for the kind of customization you need!

**Comment by Nosferican** [comment]
The rootlike structure seems similar to the native `Expr` which I think works pretty well. I think we all agree we need a better terms representation moving forward. The Julia 0.6 version of my package uses that approach of a `buildmodel(df, @model(formula = y ~ x, estimator = OLS)) ` which uses an intermediate `ModelBuilder` struct which I then use to build `MyStruct`. I wasn't sure how did the `ModelBuilder` adapts to each package (ie.e, type of the model by the call) and envisioned as basically just a `Dict{Symbol,Symbol}` to pass to the function.

A few comments:

- Sometimes is useful to have the formula per se... for example, I keep a string representation of the formula input that goes with `StatsBase.coeftable`. Thus, keeping a copy of the formula sometimes is beneficial instead of just replacing it.
- Some transformations are performed at the datasource before building the model. For example, a simple issue would be complete cases for all variables. In my case, sometimes I need to perform first difference or group means for which some are easier to do with `AbstractDataFrames` than with matrices.
- Don't know if anyone else uses this approach, but for some variables I get `Vector{Vector{Int64}}` which are used for operations by clusters/groups rather than getting a `Matrix`. When there are interactions I still need to make the model matrix and then get the indices for the levels... If there were an internal contrasts for that output, it would be appreciated.

That said, with the new more flexible `StatsBase.@formula` / `StatsBase.Formula` I use a single function to parse the components and another macro to capture keyword arguments (e.g., this is the approach in `Optim.options()`.
```
Representation, Exogenous, Endogenous, Instruments, FixedEffects = formulaparser(formula)
function formulaparser(formula::StatsModels.Formula)
    formula = copy(formula)
    representation = string(formula)
    response = formula.lhs
    formula = formula.rhs
    if formula.args[1] == :(~)
        fixedeffects = StatsModels.Formula(response, formula.args[3])
        formula = formula.args[2]
    else
        fixedeffects = StatsModels.Formula(response, zero(Int64))
    end
    instrumental = false
    for idx ∈ eachindex(formula.args)
        if isa(formula.args[idx], Expr)
            if formula.args[idx].args[1] == :(~)
                endogenous = StatsModels.Formula(response, formula.args[idx].args[2])
                instruments = StatsModels.Formula(response, formula.args[idx].args[3])
                instrumental = true
                deleteat!(formula.args, idx)
                break
            end
        end
    end
    exogenous = StatsModels.Formula(response, formula)
    if !instrumental
        endogenous = StatsModels.Formula(response, zero(Int64))
        instruments = StatsModels.Formula(response, zero(Int64))
    end
    return (representation, exogenous, endogenous, instruments, fixedeffects)
end
```
A common issue with either approach is that it would need to call `ModelFrame` on each of the formulas... A `ModelFrame` that accepted multiple formulas and contrasts would be nice to have. For example, one that returns a `Vector{ModelMatrix}`.

**Comment by Nosferican** [comment]
As for API, I agree with @matthieugomez that standardizing the way to use statistical models is the way to go. As I am unclear on how to dispatch the `ModelBuilder`, I am using this format for now...
```
modelspecs = MyPkg.@model(kw = value, kw = value,...)
model = buildmodel(df::DataFrames.AbstractDataFrame,
    modelspecs::StatsModels.AbstractModelSpecs)
coeftable(model::StatsBase.RegressionModel)
```
In StatsModels the only changes needed are:
```
abstract type AbstractModelSpecs end
buildmodel(df::DataFrames.AbstractDataFrame,
    modelspecs::AbstractModelSpecs) =
    error("buildmodel is not defined for $(typeof(obj)).")
```
with this syntax the `MyPkg.@model` is acting as the main command in Stata. It specifies what the regression model is and allows for the arguments to match the command. For example, Hausman-Taylor would accept `response + exogenous + (endogenous ~ )` vs 2SLS which would require `response + exogenous + (endogenous ~ instruments)`. It is flexible as the actual `modelbuilder` can be dispatched on a captured arguments (e.g., `estimator = logit`).

**Comment by Nosferican** [comment]
After some talk a few days back on Slack, apparently there is an undesirable trend of using macros rather than direct builders. After that discussion I tried a different approach which seems to be an improvement.
```
model = MyPkgModel(formula::StatsModels.Formula, data::DataFrames.AbstractDataFrame;
                                      options::Dict{Symbol,Any} = Dict{Symbol,Any}())
```
where I defined
```
function parsearg(obj::Expr)
    args = obj.args
    @assert (obj.head == :(=)) & (length(args) == 2) "Model options should be especified using the following syntax: Argument = Value"
    return (args[1], args[2])
end
macro options(args...)
    Dict{Symbol,Any}(parsearg.(args))
end
```
An example is,
```
model = EconometricsModel(@formula(y ~ x1 + x2 + (z1 ~ z2) + (absorb = fe1) + (cluster = fe1)),
                                                 data,
                                                 options = @options(estimator = FD, λ = 0.1)
coeftable(model)
```

**Comment by kleinschmidt** [comment]
The options macro isn't necessary here; you can just define a function that splats keyword arguments into a dict: `options(kw...) = Dict(kw)`.

I'm afraid this discussion has gotten pretty far from where it started.  Standardizing on a way of specifying (eagerly-evaluated) options for building a model might be useful, but that's a separate issue from how best to handle extracting elements from a tabular data source based on multiple arguments to a model constructor.

Also, what I'm imagining is using a macro as a convenience so you don't have to type `@formula` for every single argument (which would get tedious if you want to specify a lot of tiny one-sided formulae, which is my preferred solution for what @matthieugomez requested for things like weights, standard errors, etc.).  That would just lower to a constructor for a `ModelBuilder{ModelType}` object which has default behavior that can be overloaded by packages that need to intercept the formulae before they're interpreted.  There would be nothing stopping someone from writing out the more verbose literal form.

**Comment by Nosferican** [comment]
The idea of the macro is so people can use `val` instead of `:val` for symbol values.

For the many components formula I just use `response ~ exogenous + (endogenous ~ instruments) + (keyword = onesidedformula)` where keyword can be `absorb`, `cluster`, `weights`, etc. This wasn't possible before, but is feasible now. The default formula parser could just return a `Dict{Symbol, StatsModels.Formula}` for the `keyword = onesidedformula` components. The `ModelBuilder` could return a `Dict{Symbol,Matrix{Float64}}` as model matrixes for each of the components. An alternative procedure could be the default for specific `keyword` for example, `absorb` or `cluster` could use #37.

**Comment by kleinschmidt** [comment]
```julia
julia> f(; kw...) = Dict(kw)
f (generic function with 1 method)

julia> f(x=1)
Dict{Symbol,Int64} with 1 entry:
  :x => 1
```

**Comment by Nosferican** [comment]
```
f(estimator = FD, vce = OLS)
UndefVarError: FD not defined
```
```
f(estimator = :FD, vce = :OLS)
Dict{Symbol,Symbol} with 2 entries:
  :estimator => :FD
  :vce       => :OLS
```
```
@options(estimator = FD, vce = OLS)
Dict{Symbol,Any} with 2 entries:
  :estimator => :FD
  :vce       => :OLS
```

**Comment by kleinschmidt** [comment]
Ah I misunderstood what you meant by symbols.  This actually strikes me as a better example of "macro-creep" than having a multi-formula `@model` macro (it's pretty idiomatic to use symbols as options like this; or else to define an Enum type...)

**Comment by kleinschmidt** [comment]
@lbittarello do you have an example? I'm having trouble picturing this. That sounds like a case where a macro might be appropriate

**Comment by lbittarello** [comment]
Suppose that you want to run a linear instrumental regression. @Nosferican would write:

    ModelFrame(@formula(y ~ (t ~ z) + x), data)

I write:

    Microdata(data, response = "y", treatment = "t", instrument = "z", control = "x")

`Microdata` creates a model matrix and a dictionary. The dictionary maps each variable set into the columns of the model matrix. The regression goes:

    Y = getvector(microdata, :response)
    X = getmatrix(microdata, :treatment, :control)
    Z = getmatrix(microdata, :instrument, :control)
    β = (Z' * X) \ (Z' * Y)

I guess that `Microdata` is more verbose. But it's flexible. Suppose that you're implementing a semiparametric single-index model. You would ask the user for a `response`, `discrete` controls and `continuous` controls. No special syntax would be necessary.

**Comment by lbittarello** [comment]
It's useful too when models interpret components differently. For example, linear regression in Stata goes `regress y t x`, while inverse probability weighting is `teffects ipw (y) (t x)`. `Microdata` would use the same syntax for both.

**Comment by Nosferican** [comment]
That can be managed as well à la Stata @formula(y ~ x1 + x2 + (z1 ~ 0)) would create the same `Dict{Symbol,Matrix{Float64}}`, but if the estimator uses only endogenous variables and no additional instruments it can just use the model matrices differently to construct the modelmatrix and any transformations.

An additional thought is that sometimes those components are better served with a different representation than a `Matrix{Float64}`. For example, clusters are better as `Vector{Vector{Vector{Int64}}}` in the form of `Clusters[Dimension[Group[Observations]]]` (see #37). Maybe two outputs `Dict{Symbol,Matrix{Float64}}` and `Dict{Symbol,Vector{Vector{Vector{Int64}}}}`?

**Comment by lbittarello** [comment]
I agree that clusters are better in a separate matrix. Covariance options currently receive a different treatment. It would be easy to designate other special keywords.

**Comment by Nosferican** [comment]
A few tools I have been developing such as [VCE](https://github.com/JuliaEconometrics/VarianceCovarianceEstimators.jl) and various routines in [EconUtils](https://github.com/JuliaEconometrics/EconUtils.jl) use a vector for the response, matrices for all others, but fixed effects to be absorbed and clusters for variance covariance estimation. The last two use the `Vector{Vector{Vector{Int64}}}}` form.

**Comment by pdeffebach** [comment]
There was some discussion about this in a GLM issue [here](https://github.com/JuliaStats/GLM.jl/issues/255#issuecomment-422530778). Linking here because I think that `@formula(y ~ x, cluster(z))` or something similar, and have a ModelFrame object record that, would be very helpfu.
